api/server/server.go

```diff
@@ -99,12 +100,8 @@ func (s *Server) ServeApi(protoAddrs []string) error {
 				chErrors <- err
 				return
 			}
-			s.eng.OnShutdown(func() {
-				if err := srv.Close(); err != nil {
-					logrus.Error(err)
-				}
-			})
-			if err = srv.Serve(); err != nil && strings.Contains(err.Error(), "use of closed network connection") {
+			s.servers = append(s.servers, srv)
+			if err := srv.Serve(); err != nil && strings.Contains(err.Error(), "use of closed network connection") {
```

Comment:
I think this whole function should be rewritten:
* you need to collect errors from you need to collect errors from serving goroutines before return
* I think it is better first to collect pairs proto/addr and create all servers before serving in goroutines, which allow you to catch errors before any serve and avoid race in appending to `s.servers`. So, just reduce goroutine code to `srv.Serve`.